### PR TITLE
Added color option to waveform generation

### PR DESCRIPTION
**Pull request by iganev**
| Q                  | A
| ------------------ | ---
| Bug fix?           | no
| New feature?       | yes
| BC breaks?         | no
| Deprecations?      | no
| License            | MIT

#### What's in this PR?

Added the ability to define color palette for the waveform. The default color is black.

#### Why?

By default ffmpeg uses a mixture of colors (mostly orange) that result in an image that is very difficult for post-processing and downright ugly.

#### Example Usage

```php
$audio->waveform(1280, 200, ['#00FF00']);
//or
$waveform = $audio->waveform(1280, 200);
$waveform->setColors(['#FFFFFF']);
~~~

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -165,7 +165,7 @@ documentation below for more information.
 The ouput file MUST use the PNG extension.
 
 ```php
-$waveform = $audio->waveform(640, 120);
+$waveform = $audio->waveform(640, 120, ['#00FF00']);
```
Comment (jens1o) [inline README.md]: Could you describe the third argument, please?

File: src/FFMpeg/Media/Waveform.php [hunk 2]
```diff
@@ -66,6 +75,50 @@ public function addFilter(WaveformFilterInterface $filter)
...
+     */
+    public function setColors(array $colors)
+    {
+        foreach ($colors as $row => $value)
+        {
+            if ($value{0} != '#' || strlen($value) != 7)
+            {
+                //invalid color
+                unset($colors[$row]);
+            }
```
Comment (jens1o) [inline src/FFMpeg/Media/Waveform.php]: please use a regex instead.

File: src/FFMpeg/Media/Waveform.php [hunk 3]
```diff
@@ -66,6 +75,50 @@ public function addFilter(WaveformFilterInterface $filter)
     }
 
     /**
+     * Parameter should be an array containing at least one valid color represented as a HTML color string. For
+     * example #FFFFFF or #000000. By default the color is set to black. Keep in mind that if you save the waveform
+     * as jpg file, it will appear completely black and to avoid this you can set the waveform color to white (#FFFFFF).
+     * Saving waveforms to png is strongly suggested.
```
Comment (jens1o) [inline src/FFMpeg/Media/Waveform.php]: Please insert a new paragraph before this line.

File: src/FFMpeg/Media/Waveform.php [hunk 4]
```diff
@@ -66,6 +75,50 @@ public function addFilter(WaveformFilterInterface $filter)
...
+        {
+            $this->colors = $colors;
+        }
+    }
+
+    /**
+     * Returns an array of colors that will be passed to ffmpeg to use for waveform generation. Colors are applied ONLY
+     * to the waveform. Background cannot be controlled that easily and it is probably easier to save the waveform
+     * as a transparent png file and then add background of choice.
+     * @return array
```
Comment (jens1o) [inline src/FFMpeg/Media/Waveform.php]: new line before this line please

File: src/FFMpeg/Media/Waveform.php [hunk 5]
```diff
@@ -66,6 +75,50 @@ public function addFilter(WaveformFilterInterface $filter)
...
+     */
+    public function getColors()
+    {
+        return $this->colors;
+    }
+
+    /**
+     * Compiles the selected colors into a string, using a pipe separator.
+     * @return string
```
Comment (jens1o) [inline src/FFMpeg/Media/Waveform.php]: new line please

**Comment by jens1o** [comment]
Could you fix the tests please? When you need some help, I'm here to help you.

File: README.md (hunk 1 above)
Comment (iganev) [reply README.md]: Done

File: src/FFMpeg/Media/Waveform.php (hunk 3 above)
Comment (iganev) [reply src/FFMpeg/Media/Waveform.php]: Done

**Comment by iganev** [comment]
Sorry I was unaware that I broke the tests after adding the -y option. Totally forgot to check them.

File: src/FFMpeg/Media/Waveform.php [hunk 6]
```diff
@@ -66,6 +75,53 @@ public function addFilter(WaveformFilterInterface $filter)
...
+     * @param array $colors
+     */
+    public function setColors(array $colors)
+    {
+        foreach ($colors as $row => $value)
+        {
+            if (!preg_match('/^#(?:[0-9a-fA-F]{6})$/', $value))
+            {
+                //invalid color
+                unset($colors[$row]);
```
Comment (jens1o) [inline src/FFMpeg/Media/Waveform.php]: I'm not sure. I think this should throw an `InvalidArgumentException` instead... :c

File: README.md (hunk 1 above)
Comment (jens1o) [reply README.md]: Did you pushed it? I don't see it unfortunately. :(

Comment (iganev) [reply README.md]: There was a bit of misunderstanding here. I though you mean to update the docblock section of the method. I have updated the README.md now. :)

File: src/FFMpeg/Media/Waveform.php (hunk 6 above)
Comment (iganev) [reply src/FFMpeg/Media/Waveform.php]: That sounds reasonable.

File: README.md [hunk 7]
```diff
@@ -159,13 +159,13 @@ You can generate a waveform of an audio file using the `FFMpeg\Media\Audio::wave
...
 This code returns a `FFMpeg\Media\Waveform` instance.
-You can optionally pass dimensions as arguments, see dedicated
+You can optionally pass dimensions as the first two arguments and an array of hex string colors for ffmpeg to use for the waveform, see dedicated
 documentation below for more information.
 
 The ouput file MUST use the PNG extension.
 
 ```php
-$waveform = $audio->waveform(640, 120, ['#00FF00']);
+$waveform = $audio->waveform(640, 120, ['#00FF00/']);
```
Comment (jens1o) [inline README.md]: is that slash placed by purpose?

Comment (iganev) [reply README.md]: Not that I know of.. :D I've asked the cat, but she denies all allegations.

File: README.md (hunk 1 above)
Comment (jens1o) [reply README.md]: This also needs to use the `array()` syntax. :(

File: src/FFMpeg/Media/Waveform.php [hunk 8]
```diff
@@ -20,17 +21,26 @@
...
+    protected $width;
+    protected $height;
+
+    /**
+     * @var array
+     */
+    protected $colors;
 
-    public function __construct(Audio $audio, FFMpegDriver $driver, FFProbe $ffprobe, $width, $height)
+    public function __construct(Audio $audio, FFMpegDriver $driver, FFProbe $ffprobe, $width, $height, $colors = [self::DEFAULT_COLOR])
```
[the remark this answers is not in the document]
Comment (iganev) [reply src/FFMpeg/Media/Waveform.php]: I have updated:
`README.md`
`src/FFMpeg/Media/Audio.php`
`src/FFMpeg/Media/Waveform.php`

**Comment by jens1o** [comment]
Thanks!

**Comment by iganev** [comment]
Glad I can help! :)

**Comment by jens1o** [comment]
Release is on its way ;)

**Comment by jens1o** [comment]
It's released. Thanks again! :)

**Pull request merged**